rabbitmq/retryable_producer.go

```diff
@@ -0,0 +1,169 @@
+	HealthCheckFactor  int
+	BackoffConfig      *backoff.Config
+	RabbitClientConfig *ClientConfig
+}
+
+func NewRetryableProducer(
+	newClientFactory func(ctx context.Context, config *ClientConfig) (RabbitMQClientInterface, error),
+	config RetryableProducerConfig,
+	logger logger.StructuredLogger,
+	metric Metric,
+) *RetryableProducer {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	retryableProducer := &RetryableProducer{
+		logger:        logger,
+		metric:        metric,
+		clientFactory: newClientFactory,
+		config:        config,
+		cancel:        cancel,
+	}
+
+	go retryableProducer.initProducer(ctx)
+
+	return retryableProducer
+}
+
+func (p *RetryableProducer) Publish(
+	exchange,
+	key string,
+	mandatory,
+	immediate bool,
+	expiration string,
+	body []byte,
+	args amqp.Table,
+) error {
+	p.mu.RLock()
+	producer := p.producer
+	p.mu.RUnlock()
+
+	if producer == nil {
+		return stacktrace.NewError("RabbitMQ Producer client not connected")
+	}
+
+	err := producer.Publish(exchange, key, mandatory, immediate, expiration, body, args)
+	if err != nil {
+		return stacktrace.Propagate(err, "failed to publish RMQ message")
+	}
+
+	return nil
+}
+
+func (p *RetryableProducer) newProducer(ctx context.Context) (*Producer, error) {
+	if ctx.Err() != nil {
+		p.logger.Info("received context cancel")
+		return nil, nil
+	}
+
+	client, err := p.clientFactory(ctx, p.config.RabbitClientConfig)
+	if err != nil {
+		return nil, stacktrace.Propagate(err, "RabbitMQ Failed to init client")
+	}
+
+	producer, err := NewProducer(client, p.logger, p.metric)
+	if err != nil {
+		connCloseErr := client.Close()
+		p.logger.Error("cannot close RabbitMQ client connection", zap.Error(connCloseErr))
+		return nil, stacktrace.Propagate(err, "RabbitMQ Failed to create new producer")
+	}
+
+	return producer, nil
+}
+
+func (p *RetryableProducer) initProducer(ctx context.Context) {
+	producerBackoff := backoff.NewBackoff(p.config.BackoffConfig)
+	currentRetryAttempts := 0
+
+	for {
+		producer, err := p.newProducer(ctx)
+		if err != nil {
+			p.logger.Error("producer connection failed with error", zap.Error(err))
+
+			if p.config.MaxRetryAttempts != 0 && currentRetryAttempts > p.config.MaxRetryAttempts {
+				p.logger.Info("retry attempts exceeded")
+				return
+			}
+
+			backoffDuration := producerBackoff.Next()
+
+			select {
+			case <-ctx.Done():
+				p.logger.Info("received context cancel")
+				return
+			case <-time.After(backoffDuration):
+				continue
+			}
+		}
+
+		p.mu.Lock()
+		p.producer = producer
+		p.mu.Unlock()
+
+		p.logger.Info("producer connected to RabbitMQ")
+
+		select {
+		case <-ctx.Done():
+			p.logger.Info("received shut down signal")
+			p.Close()
+			return
+		case <-producer.closeCh:
+			p.logger.Info("RabbitMQ Producer Client closed the connection, trying to reconnect")
+		}
+	}
+}
+
+func (p *RetryableProducer) Close() {
+	var once sync.Once
```

Comment:
The once var must be a private field not a local variable. Multiple calls of `Close()` must hit the same `once` instance.

Reply:
Actually, the producer must be closed by the initProducer routine. When it receives ctx.Done it can close the producer and exit. 

In which case here the `Close` method do not need `once` and can just call `p.cancel()`.

Reply:
Yeah this simplifies things a bit. Pushed the changes.